packages/squiggle-lang/src/rescript/Reducer/Reducer_MathJs/Reducer_MathJs_Parse.res

```diff
@@ -0,0 +1,158 @@
+}
+type accessorNode = {
+  ...node,
+  "object": node,
+  "index": indexNode
+}
+type operatorNode = {
+  ...functionNode,
+  "op": string,
+}
+
+//parenthesisNode
+type parenthesisNode = {
+  ...node,
+  "content": node
+}
+//rangeNode
+//relationalNode
+type symbolNode = {
+  ...node,
+  "name": string
+}
+
+external castAccessorNode: node => accessorNode = "%identity"
+external castArrayNode: node => arrayNode = "%identity"
+external castConstantNode: node => constantNode = "%identity"
+external castFunctionNode: node => functionNode = "%identity"
+external castIndexNode: node => indexNode = "%identity"
+external castObjectNode: node => objectNode = "%identity"
+external castOperatorNode: node => operatorNode = "%identity"
+external castOperatorNodeToFunctionNode: operatorNode => functionNode = "%identity"
+external castParenthesisNode: node => parenthesisNode = "%identity"
+external castSymbolNode: node => symbolNode = "%identity"
+
+/*
+  MathJs Parser
+*/
+@module("mathjs") external parse__: string => node = "parse"
+
+let parse = (expr: string): result<node, reducerError> =>
+  try {
+    Ok(parse__(expr))
+  } catch {
+  | Js.Exn.Error(obj) =>
+    RerrJs(Js.Exn.message(obj), Js.Exn.name(obj))->Error
+  }
+
+type mjNode =
+  | MjAccessorNode(accessorNode)
+  | MjArrayNode(arrayNode)
+  | MjConstantNode(constantNode)
+  | MjFunctionNode(functionNode)
+  | MjIndexNode(indexNode)
+  | MjObjectNode(objectNode)
+  | MjOperatorNode(operatorNode)
+  | MjParenthesisNode(parenthesisNode)
+  | MjSymbolNode(symbolNode)
+
+let castNodeType = (node: node) => switch node["type"] {
+  | "AccessorNode" => node -> castAccessorNode -> MjAccessorNode -> Ok
+  | "ArrayNode" => node -> castArrayNode -> MjArrayNode -> Ok
+  | "ConstantNode" => node -> castConstantNode -> MjConstantNode -> Ok
+  | "FunctionNode" => node -> castFunctionNode -> MjFunctionNode -> Ok
+  | "IndexNode" => node -> castIndexNode -> MjIndexNode -> Ok
+  | "ObjectNode" => node -> castObjectNode -> MjObjectNode -> Ok
+  | "OperatorNode" => node -> castOperatorNode -> MjOperatorNode -> Ok
+  | "ParenthesisNode" => node -> castParenthesisNode -> MjParenthesisNode -> Ok
+  | "SymbolNode" => node -> castSymbolNode -> MjSymbolNode -> Ok
+  | _ => Rerr.RerrTodo(`Argg, unhandled MathJsNode: ${node["type"]}`)-> Error
+}
+
+let rec show = (mjNode: mjNode): string => {
+  let showValue = (a: 'a): string => if (Js.typeof(a) == "string") {
+    `'${Js.String.make(a)}'`
+  } else {
+    Js.String.make(a)
+  }
+
+  let showNodeArray = (nodeArray: array<node>): string =>
+    nodeArray
+    -> Belt.Array.map( a => showMathJsNode(a) )
+    -> AE.interperse(", ")
+    -> Js.String.concatMany("")
+
+  let showFunctionNode = (fnode: functionNode): string =>
+    `${fnode["fn"]}(${fnode["args"]->showNodeArray})`
+
+  let showObjectEntry = ( (key: string, value: node) ): string =>
+    `${key}: ${value->showMathJsNode}`
+
+  let showObjectNode = (oNode: objectNode): string =>
```

Comment:
Could you please explain to me this pattern of writing rescript code in strings? is this gonna be like python's notion of `eval`? 

Why did `showValue` put it in a conditional but the others do not?

Reply:
let showValue = (a: 'a): string => if (Js.typeof(a) == "string") {
    `'${Js.String.make(a)}'`
  } else {
    Js.String.make(a)
  }
  
does simply print the strings quoted so that we get
    hello (string) prints 'hello'
    hello (symbol) prints hello

Reply:
thanks!

Reply:
You are looking at the internals of let rec show = (mjNode: mjNode): string 
This is a function to print MathJs node trees as they are. So that I can debug the unit tests on converting mathjs node tree to what we need.  It has nothing to do with the actual evaluation.

It is conditional becase we want strings to be printed quoted so that we know they are strings while debugging